File: problems/isomorphic-strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char, char> sump;
        unordered_map<char, char> tump;
        char ch = 'a';
        for(int i = 0; i < s.size(); i++){
            if(sump.find(s[i]) != sump.end()){
                if(tump.find(t[i]) == tump.end()) 
                    return false; 
                else {
                    s[i] = sump[s[i]];
                    t[i] = tump[t[i]];
                }
            } else {
                if(tump.find(t[i]) != tump.end()) return false;
                sump[s[i]] = ch;
                tump[t[i]] = ch;
                s[i] = ch;
                t[i] = ch++;
            }
            
        }
        
        return s == t;
    }
};


/*
    Complexity analysis:
    Time: O(n), where `n` is the length of the given strings, as we need to scan every char of the two strings.
    Space: O(1), since in worst case the the maps will hold 26 characters each.
*/

File: problems/isomorphic-strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        int n = s.length();
        int[] map1 = new int[256];
        int[] map2 = new int[256];
        
        Arrays.fill(map1, -1);
        Arrays.fill(map2, -1);
        
        for(int i = 0; i < n; i++){
            char chs = s.charAt(i);
            char cht = t.charAt(i);
            if(map1[chs] != map2[cht]) return false;
            
            map1[chs] = map2[cht] = i;
        }
        
        return true;
    }
}


/*
    Complexity analysis:
    Time: O(n), where `n` is the length of the given strings, as we need to scan every char of the two strings.
    Space: O(1), since in worst case the the maps will hold 26 characters each.
*/
